crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+    fn is_supported(&self) -> bool {
+        match self {
+            Identity | Hardware | Contact | Information => true,
+            Custodian => false,
+        }
+    }
+}
+
+/// A crate internal helper builder which assigns factors to
+/// Recovery and Confirmation roles based on the heuristic
+/// laid out in ["Automatic Security Shield Construction" document][doc]
+///
+/// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+pub(crate) struct AutomaticShieldBuilder {
+    /// Only used for testing purposes, feel free to remove.
+    stats_for_testing: AutoBuildOutcomeForTesting,
+
+    /// While we assign factors to Recovery and Confirmation roles, we remove
+    /// them from this set.
+    remaining_available_factors: IndexSet<FactorSource>,
+
+    /// The factors assigned to each role, including the factors originally
+    /// set for the primary role.
+    proto_matrix: ProtoMatrix,
+}
+
+impl SecurityShieldBuilder {
+    /// Assigns the factors to the Recovery and Confirmation roles according to the heuristic
+    /// laid out in ["Automatic Security Shield Construction" document][doc].
+    ///
+    /// The `all_factors_in_profile` should contain all factors that are available for the user,
+    /// i.e. from Profile, and SHOULD contain the factors that the user has preselected for the
+    /// Primary role.
+    ///
+    /// # Throws
+    /// Throws if the primary role is invalid.
+    /// Throws if the primary override factors are not empty.
+    /// Throws if the primary factors are not in the profile.
+    /// Throws if the prerequisites are not met.
+    /// Throws if the shield is invalid after auto assignment - which should not happen, can be considered programmer error!
+    ///
+    /// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+    pub fn auto_assign_factors_to_recovery_and_confirmation_based_on_primary(
+        &self,
+        all_factors_in_profile: IndexSet<FactorSource>,
+    ) -> Result<AutoBuildOutcomeForTesting>
+/* Feel free to replace `AutoBuildOutcomeForTesting` return type if you need anything else, I had Unit, so might as well make testing easier by returning this type. */
+    {
+        if let Some(invalid_reason) = self.validate_primary_role() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: format!(
+                    "Primary role is not valid: {:?}",
+                    invalid_reason
+                ),
+            });
+        }
+
+        if !self.get_primary_override_factors().is_empty() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Primary override factors not allowed when preselecting factors for Recovery and Confirmation".to_string(),
+            });
```

Comment:
Can you elaborate why there should be no override factors in primary?

Reply:
All of Matts SecurityShield building uses Threshold list for Primary.